### ISSUE TITLE
[JS] Anonymous vs closure functions

### ISSUE THREAD

**Issue by phuocng** [phuocng/this-vs-that#283]

**Comment by MkDev11** [comment]
Anonymous functions and closure functions are both concepts in JavaScript related to the way functions are defined and how they capture and interact with variables in their surrounding scope. Let's explore each of these concepts:

**Anonymous Functions:**

- **Definition:** An anonymous function is a function without a name. It is defined inline and can be assigned to a variable or passed as an argument to other functions.

- **Example:**
  ```javascript
  const add = function (a, b) {
    return a + b;
  };

  const result = add(3, 5);
  ```

- **Usage:** Anonymous functions are often used when you need a function for a specific purpose without the need to give it a name. They are commonly used in callback functions, event handlers, and as arguments for array methods like `map`, `filter`, and `reduce`.

**Closure Functions:**

- **Definition:** A closure function is a function that "closes over" variables from its outer scope, preserving access to those variables even after the outer function has finished executing.

- **Example:**
  ```javascript
  function outer() {
    const x = 10;

    function inner(y) {
      return x + y;
    }

    return inner;
  }

  const closureFunction = outer();
  const result = closureFunction(5); // result is 15
  ```

- **Usage:** Closure functions are commonly used when you need to encapsulate state and behavior in a way that keeps data private and provides controlled access to it. They are often used in scenarios like data hiding and the module pattern in JavaScript.

**Key Differences:**

1. **Naming:** Anonymous functions have no names or identifiers, while closure functions have names.

2. **Scope:** Both types of functions can capture and access variables from their outer scope. However, closure functions are specifically known for their ability to create closures, which allows them to retain access to those variables even after the outer function has completed execution.

3. **Use Cases:** Anonymous functions are typically used for short-lived, one-off tasks like callbacks. Closure functions are used to create self-contained and stateful modules or to capture and maintain state across multiple function calls.

In summary, anonymous functions are a common and convenient way to define functions without naming them, often used in callback scenarios. Closure functions, on the other hand, are functions that capture and preserve access to variables from their surrounding scope, making them a powerful tool for encapsulating state and behavior. While they are different concepts, a closure function can also be an anonymous function if it's defined without a name.

**Comment by phuocng** [comment]
Done: https://phuoc.ng/collection/this-vs-that/anonymous-vs-closure-functions/
